sympy/matrices/determinant.py

```diff
@@ -410,10 +409,30 @@ def _charpoly(M, x='lambda', simplify=_simplify):
             m = m * (x - simplify(i))
         return PurePoly(m, x)
 
-    berk_vector = _berkowitz_vector(M)
-    x = uniquely_named_symbol(x, berk_vector, modify=lambda s: '_' + s)
+    xdum = Dummy('t')
+    DOM = DomainMatrix.from_Matrix(M, radicals_limit=1)
```

Comment:
Is `1` a good choice for `radicals_limit`? Have you tried with matrices having more than one radical?

Reply:
I blatantly assumed that only `1` will be a good choice, So, I ran some tests for `3*3` and `5*5` matrices.
I found `radicals_limit=5` is the optimum choice, as for `radicals_limit=6`, the `DomainMatrix.from_Matrix` hangs.

Reply:
Choosing only one is a conservative choice but maybe that's okay. It means that `primitive_element` is never even called. Choosing a higher value means that we need to predict when `primitive_element` will be slow.

I expect that the optimal choice will also depend on the degree of the radical e.g. if you use `cbrt(2)`, `cbrt(3)` or `3**(S(1)/10)` etc then maybe a lower value of radical limit would be needed.

It might also be preferable to use a different test rather than just radicals because the same sort of things can happen with other algebraic expressions e.g.:
```python
In [5]: vals = [exp(I*pi*n/13)+1 for n in range(13)]

In [6]: primitive_element(vals, ex=True)
^C---------------------------------------------------------------------------
KeyboardInterrupt
```

Reply:
> I expect that the optimal choice will also depend on the degree of the radical e.g. if you use `cbrt(2)`, `cbrt(3)` or `3**(S(1)/10)` etc then maybe a lower value of radical limit would be needed.
> 
Indeed, for `3**(S(1)/10)`, it couldn't hold on to even `2` radicals.
It will probably be best to keep it `1` for now.

There is a suggestion, can we make create some function to determine the `radicals_limit` for different matrices, better to say for `items_sympy` in `DomainMatrix.get_Domain`. 
I mean is there some logic to it, as to why for `sqrt(a)`, it holds up to 5, while for `cbrt(a)`, it is just 3.
Thanks.

Reply:
You can try the `minpoly` and `primitive_element` functions yourself to see what's going on. It's really to do with the degree of the polynomials involved e.g.:
```python
In [19]: minpoly(sqrt(2))
Out[19]: 
 2    
x  - 2

In [20]: minpoly(3**(S(1)/10))
Out[20]: 
 10    
x   - 3

In [22]: primitive_element([sqrt(2), sqrt(3)])
Out[22]: 
⎛ 4       2            ⎞
⎝x  - 10⋅x  + 1, [1, 1]⎠

In [23]: primitive_element([sqrt(2), cbrt(3)])
Out[23]: 
⎛ 6      4      3       2                   ⎞
⎝x  - 6⋅x  - 6⋅x  + 12⋅x  - 36⋅x + 1, [1, 1]⎠

In [24]: primitive_element([sqrt(2), 2**(S(1)/10)])
Out[24]: 
⎛ 10       8       6        4             ⎞
⎝x   - 10⋅x  + 40⋅x  - 100⋅x  - 50, [1, 1]⎠
```
In general `primitive_element` can produce a new algebraic element with a degree that is the product of the degrees of its inputs. For square roots that means that with `n` radicals the degree can be `2**n`. With 10th roots it can be `10**n`.

Reply:
It might be better to have a limit in the `primitive_element` function. The slow part is factoring a large polynomial over an extension field so the function could check before doing that and have some limit where it just gives up. Then `construct_domain` could fall back to `EX`.

Reply:
> so the function could check before doing that and have some limit where it just gives up. Then `construct_domain` could fall back to `EX`.
>
@oscarbenjamin, Should the check be some kind of time limit, maybe using `signal.SIGALRM`?
Thanks.

Reply:
The check should be something like the degree of the large polynomial and the extension before calling factor (if that's the slow part).

Reply:
1. The timings are not linear to the degree of polynomials.
For e.g. 
`primitive_element([2**(S(1)/10), cbrt(2), cbrt(3), cbrt(5)])` takes `34.543` seconds and the evaluated polynomial is of degree `270`, on the other hand,  `primitive_element([cbrt(2), cbrt(3), cbrt(5), cbrt(7)])` took `431.039` seconds while the evaluted polynomial is of degree just `81`.
Here we can say that when radicals of the same type consume more time than a variety of radicals, and this is seen in other cases.

2. The `minimal_polynomial` is very fast in general for simple radicals like `sqrt(a)`, `cbrt(a)`, `tan(a*pi)` etc. But if we complicate the radical like `cos(13*pi/45)*sin(17*pi/43)`, or even more complicating, it slows down a lot.

3. In general, for simple radicals(simple is not a good word to define, I mean to say radicals who have direct methods to evaluate), we can blame `factor` for slowing down the construction.

Clearly, the density of each type of radicals matters more than the degree of the Polynomial. 
If the density of the particular radical increases, it affects the time elapsed than the degree of polynomials.

This is enough evidence to conclude that we should be putting a limit on the maximum density of the radicals that can be present, and accordingly abort the calling of factor and the domain could be `EX`.

The question is how should we be framing it, the `limit`.
For the `Pow` objects, I already went through the limits, like `sqrt(a)-5`, `cbrt(a)-3`, `3**(S(1)/10)-1`.
Similarly for other algebraic radicals, we can find the limit, Perhaps we can set the limit to 1, and vary accordingly if all radicals are `sqrt`...
Will have to look deeper into the idea, just sharing some pointers.

@oscarbenjamin, kindly share your opinion,
Thanks a lot.

Reply:
I think that the best approach would be to build the limits into the `primitive_element` and `minpoly` functions themselves so that you can do something like
```python
minpoly(expr, limit=...)
```
Then internally those routines can stop before attempting some operation that would be slow like factoring a large poly.